File: android/src/main/kotlin/io/getimpulse/player/plugin/core/PluginNativeViewFactory.kt
package io.getimpulse.player.plugin.core

import android.content.Context
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory
import io.getimpulse.player.Navigator
import io.getimpulse.player.extension.ImpulsePlayerFlutter
import io.getimpulse.player.model.PlayerDelegate
import io.getimpulse.player.plugin.ImpulsePlayerViewContainer

internal class PluginNativeViewFactory(
    private val channel: MethodChannel,
    private val navigator: Navigator,
) : PlatformViewFactory(StandardMessageCodec.INSTANCE) {

    companion object {
        private val instances = mutableMapOf<Int, ImpulsePlayerViewContainer>()

        fun get(id: Int) = instances[id]

        fun dispose(instance: ImpulsePlayerViewContainer) {
            instances.remove(instance.id)
        }
    }

    override fun create(context: Context, id: Int, args: Any?): PlatformView {
        val instance = ImpulsePlayerViewContainer(id, context, object : PlayerDelegate {
            override fun onReady() {
                PluginCallback.OnReady.execute(channel, id)
            }

            override fun onPlay() {
                PluginCallback.OnPlay.execute(channel, id)
            }

            override fun onPause() {
                PluginCallback.OnPause.execute(channel, id)
            }

            override fun onFinish() {
                PluginCallback.OnFinish.execute(channel, id)
            }

            override fun onError(message: String) {
                PluginCallback.OnError(message).execute(channel, id)
            }
        })
        ImpulsePlayerFlutter.setNavigator(instance.view, navigator)
        instances[id] = instance
        return instance
    }
}

File: ios/Classes/Core/PluginNativeViewFactory.swift
import Flutter
import ImpulsePlayer

class PluginNativeViewFactory: NSObject {
    
    private let channel: FlutterMethodChannel
    
    init(channel: FlutterMethodChannel) {
        self.channel = channel
    }
    
    private static var instances = [Int64: ImpulsePlayerViewContainer]()
    
    static func get(_ id: Int64) -> ImpulsePlayerViewContainer? {
        return instances[id]
    }
    
    static func dispose(_ instance: ImpulsePlayerViewContainer) {
        instances.removeValue(forKey: instance.id)
    }
}

extension PluginNativeViewFactory: FlutterPlatformViewFactory {
    
    func create(
         withFrame frame: CGRect,
         viewIdentifier viewId: Int64,
         arguments args: Any?
     ) -> FlutterPlatformView {
         print("factory create \(viewId)")
         let instance = ImpulsePlayerViewContainer(id: viewId, delegate: PluginPlayerDelegate(channel, viewId))
         Self.instances[viewId] = instance
         return instance
     }
}

class PluginPlayerDelegate {
    let channel: FlutterMethodChannel
    let id: Int64
    
    init(_ channel: FlutterMethodChannel, _ id: Int64) {
        self.channel = channel
        self.id = id
    }
}

extension PluginPlayerDelegate: PlayerDelegate {
    
    func onReady(_ impulsePlayerView: ImpulsePlayerView) {
        print("factory onReady")
        PluginCallback.OnReady().execute(channel, id)
    }
    
    func onPlay(_ impulsePlayerView: ImpulsePlayerView) {
        print("factory onPlay")
        PluginCallback.OnPlay().execute(channel, id)
    }
    
    func onPause(_ impulsePlayerView: ImpulsePlayerView) {
        print("factory onPause")
        PluginCallback.OnPause().execute(channel, id)
    }
    
    func onFinish(_ impulsePlayerView: ImpulsePlayerView) {
        print("factory onFinish")
        PluginCallback.OnFinish().execute(channel, id)
    }
    
    func onError(_ impulsePlayerView: ImpulsePlayerView, message: String) {
        print("factory onError")
        PluginCallback.OnError(message: message).execute(channel, id)
    }
}
